# Project 2

## imports

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

# Part 1: Data exploration

### Frequency

Frequency is the measurement of the speed of vibration in sound waves, which determines the pitch of the sound. It is the number of wave cycles that occur in one second, measured in Hertz (Hz).

### Median frequency

Is the middle value of a sorted dataset.

### Output label

The output label is used to identify which gender the voice belongs to.

# Part 2: Data preparation

In [4]:
df = pd.read_csv("voice.csv")

In [5]:
df.loc[df['label'] == 'male', 'label'] = 1
df.loc[df['label'] == 'female', 'label'] = 0

y = np.array(df['label'],dtype=int)
X = np.array(df.iloc[:, 0:20],dtype=float)

In [11]:
from sklearn.model_selection import KFold


kf = KFold(n_splits=10, shuffle=True)

for train_index, test_index in kf.split(X):
    #print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)   
    
#folds = [(X[train_index],X[test_index],y[train_index],y[test_index])for train_index, test_index in kf.split(X)]

(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2851, 20) (317, 20) (2851,) (317,)
(2852, 20) (316, 20) (2852,) (316,)
(2852, 20) (316, 20) (2852,) (316,)


# Part 3: Model construction

In [45]:
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler


pipeline1 = make_pipeline(StandardScaler(), LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial'))
#Pipeline([('clf', LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial'))])
pipeline2 = make_pipeline(StandardScaler(), svm.SVC(gamma='scale'))
#Pipeline([('SVC', svm.SVC(gamma='scale'))])
pipeline3 = make_pipeline(StandardScaler(), DecisionTreeClassifier(random_state=0))
#Pipeline([('dtc', DecisionTreeClassifier(random_state=0))])
pipeline4 = make_pipeline(StandardScaler(), KNeighborsClassifier(n_neighbors=3))
#Pipeline([('KNeigh', KNeighborsClassifier(n_neighbors=3))])

- Logistic Regression

    I think this model will perform well, because the dependtent variable is binary(male/female). 
    
    
- Support Vector Machine classifier / k-Nearest Neighbors classifier

    I think this or the k-Nearest Neighbors classifier will perform bettter than the Decision Tree classifier and almost as good as the Logistic regression.




- Decision Tree classifier

    Compared to the other classifiers, we think this we perform the worst

# Part 4: Model validation

In [37]:
from sklearn.model_selection import cross_val_score

models = [pipeline1,pipeline2,pipeline3,pipeline4]
model_name = ['Logistic regression','Support vector','Decision Tree','K-neighbours']
for index,model in enumerate(models):
    print(model_name[index])
    scores = cross_val_score(model,X, y, cv=10)
    print('Score:', scores)
    print('Mean: ',sum(scores)/len(scores))
    print('---------------------')    

Logistic regression
Score: [0.91194969 0.96855346 0.96855346 0.97484277 0.96202532 0.98734177
 0.98417722 0.9778481  0.94620253 0.98417722]
Mean:  0.9665671522967918
---------------------
Support vector
Score: [0.93081761 0.95597484 0.96855346 0.9591195  0.96835443 0.99683544
 0.98734177 0.98101266 0.91455696 0.99367089]
Mean:  0.9656237560703765
---------------------
Decision Tree
Score: [0.89622642 0.91194969 0.97484277 0.94654088 0.96518987 0.9778481
 0.99050633 0.98101266 0.90506329 0.94936709]
Mean:  0.9498547090199825
---------------------
K-neighbours
Score: [0.90566038 0.93710692 0.9591195  0.92767296 0.95886076 0.98734177
 0.98734177 0.98417722 0.89873418 0.9778481 ]
Mean:  0.9523863545896027
---------------------


## Hand in

The score shows how often the classifer predicts the value correctly. The score is a value between 0 and 1.

The scores are different due to the fact that the test and traning data changes K times. 

# Part 5: Model optimisation: scaling

We have already used StandardScaler in part 4. Because running the code without Scarler produced some warnings and to fix these warnings we needed to normalize the data.

Scaling is used to change data so it fits with some boundaries.

The mean scores : 
    
    Logistic regression - 0.9665671522967918
    Support vector - 0.9656237560703765
    Decision Tree - 0.9498547090199825
    K-neighbours - 0.9504975718493751

Our preditions matched the results quit well. Decision Tree was the worst and Logistic regression classifier was the best.

# Part 6: Manual Hyperparameter Tuning

In [52]:
rate = 0;
best_i = 0
for i in range(1,11):
    model = make_pipeline(StandardScaler(), KNeighborsClassifier(n_neighbors=i))
    scores = cross_val_score(model,X, y, cv=10)
    avg_score = sum(scores)/len(scores)
    print(i,avg_score)
    if avg_score > rate:
        rate = avg_score
        best_i = i
print('Best value for k : ',best_i)       

1 0.9441983122362869
2 0.9476833054693099
3 0.9504975718493751
4 0.9527207228723829
5 0.9523863545896027
6 0.9520838309051827
7 0.9501811161531727
8 0.9520758697555929
9 0.9508080566833852
10 0.9514449486505852
Best value for k :  4
